### DISCUSSION TITLE
posts/Go/go_redis/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#349] [Announcements]
https://www.liwenzhou.com/posts/Go/go_redis/ 

 李文周的Blog Go语言 Redis 缓存 key-value kv 存储

**Comment by heshiyingx** [comment]
go-redis的Watch似乎无效，我测试发现watch()之后更改了watch的key的值后，不会返回错误，照常执行。

**Comment by Q1mi** [comment]
> @heshiyingx
> go-redis的Watch似乎无效，我测试发现watch()之后更改了watch的key的值后，不会返回错误，照常执行。

可能是我的例子没有很清楚，你可以参照下面的例子体味这里watch的用法：当redis中某个key可能被并发的操作时，我想要在别人没有修改这个key的前提下去修改这个key的值。

```go
// 监视watch_count的值，并在值不变的前提下将其值+1
key := "watch_count"
if err := rdb.Watch(func(tx *redis.Tx) error {
	n, err := tx.Get(key).Int()
	if err != nil && err != redis.Nil {
		return err
	}
	_, err = tx.Pipelined(func(pipe redis.Pipeliner) error {
		time.Sleep(time.Second * 4) // 此时在另外一个终端窗口中使用set命令修改watch_count的值
		pipe.Set(key, n+1, 0)
		return nil
	})
	return err
}, key); err != nil {
	fmt.Printf("key发生变化，更新失败。err:%v\n", err)
} else {
	fmt.Println("更新成功")
}
```

**Comment by zhangStarV** [comment]
这个go-redis 库的poo怎么用呀七米老哥，看到newClient里有pool的设置，然后这个pool似乎也有自己的连接接口，还有GET put方法获得放回连接，，，不太明白两者怎么使用，能烦请提供一些思路或者demo么？？

**Comment by Q1mi** [comment]
> @zhangStarV
> 这个go-redis 库的poo怎么用呀七米老哥，看到newClient里有pool的设置，然后这个pool似乎也有自己的连接接口，还有GET put方法获得放回连接，，，不太明白两者怎么使用，能烦请提供一些思路或者demo么？？

Client本身就是连接池的模式，表示若干个实际的连接。
> Client is a Redis client representing a pool of zero or more underlying connections. It's safe for concurrent use by multiple goroutines.

**Comment by tianshimoyi** [comment]
七米老师，watch那个地方写错了，watch里面应该是事务(TxPipeline),不是Pipeline.
watch会在事务开始之前盯住1个或者多个变量，当事务exec时，Redis会检查关键变量自watch之后，是否被修改了（包括当前事务的客户端）。如果被改过了，则会返回null，告知客户端执行失败，这时候客户端通常会重试。

**Comment by Q1mi** [comment]
> @tianshimoyi
> 七米老师，watch那个地方写错了，watch里面应该是事务(TxPipeline),不是Pipeline.
> watch会在事务开始之前盯住1个或者多个变量，当事务exec时，Redis会检查关键变量自watch之后，是否被修改了（包括当前事务的客户端）。如果被改过了，则会返回null，告知客户端执行失败，这时候客户端通常会重试。

感谢反馈，确实应该是TxPipeline，之前笔误现已修正。

**Comment by hooting** [comment]
V8版本需要context，是漏传了吗？如`_, err = rdb.Ping().Result()`

**Comment by Jasmine456** [comment]
> @hooting
> V8版本需要context，是漏传了吗？如`_, err = rdb.Ping().Result()`

我也遇到了，可能是之前不需要，redis驱动库新增的吧
